### ISSUE TITLE
Problem with `list(sorted(...))`: `SupportsLessThanT` 

### ISSUE THREAD

**Issue by MihanixA** [python/mypy#10751]
**Bug Report**

mypy check reports an error for a code that seems to be correct 

**Brief Summary***

```python3
from typing import Dict, List, Optional


def f(row: Dict[str, str], cols: List[str]) -> List[Optional[str]]:
    return list(sorted({row.get(key) for key in cols}))
```

```bash
sample.py:5: error: Value of type variable "SupportsLessThanT" of "sorted" cannot be "Optional[str]"
```

**To Reproduce**

1. Prepare environment & create sample file

```
mkdir -p mypy-sample && cd mypy-sample
python3 -m virtualenv venv
source venv/bin/activate
pip install mypy
echo """
from typing import Dict, List, Optional


def f(row: Dict[str, str], cols: List[str]) -> List[Optional[str]]:
    return list(sorted({row.get(key) for key in cols}))
""" > sample.py
```
2. Run mypy
```
mypy sample.py -v
```

**Expected Behavior**

This code should be fine, expected
```
Success: no issues found in 1 source file
```

**Actual Behavior**

```

LOG:  Mypy Version:           0.910
LOG:  Config File:            Default
LOG:  Configured Executable:  /Users/m.a.sitnikov/devel/mypy-tests/venv/bin/python
LOG:  Current Executable:     /Users/m.a.sitnikov/devel/mypy-tests/venv/bin/python
LOG:  Cache Dir:              .mypy_cache
LOG:  Compiled:               True
LOG:  Exclude:                
LOG:  Found source:           BuildSource(path='sample.py', module='sample', has_text=False, base_dir='/Users/m.a.sitnikov/devel/mypy-tests')
LOG:  Could not load cache for sample: sample.meta.json
LOG:  Metadata not found for sample
LOG:  Parsing sample.py (sample)
LOG:  Metadata fresh for typing: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/typing.pyi
LOG:  Metadata fresh for builtins: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/builtins.pyi
LOG:  Metadata fresh for collections: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/collections/__init__.pyi
LOG:  Metadata fresh for sys: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/sys.pyi
LOG:  Metadata fresh for abc: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/abc.pyi
LOG:  Metadata fresh for types: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/types.pyi
LOG:  Metadata fresh for _typeshed: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/_typeshed/__init__.pyi
LOG:  Metadata fresh for ast: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/ast.pyi
LOG:  Metadata fresh for io: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/io.pyi
LOG:  Metadata fresh for typing_extensions: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/typing_extensions.pyi
LOG:  Metadata fresh for _collections_abc: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/_collections_abc.pyi
LOG:  Metadata fresh for importlib.abc: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/importlib/abc.pyi
LOG:  Metadata fresh for importlib.machinery: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/importlib/machinery.pyi
LOG:  Metadata fresh for array: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/array.pyi
LOG:  Metadata fresh for mmap: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/mmap.pyi
LOG:  Metadata fresh for os: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/os/__init__.pyi
LOG:  Metadata fresh for _ast: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/_ast.pyi
LOG:  Metadata fresh for codecs: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/codecs.pyi
LOG:  Metadata fresh for importlib: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/importlib/__init__.pyi
LOG:  Metadata fresh for posix: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/posix.pyi
LOG:  Metadata fresh for subprocess: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/subprocess.pyi
LOG:  Metadata fresh for os.path: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/os/path.pyi
LOG:  Metadata fresh for posixpath: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/posixpath.pyi
LOG:  Metadata fresh for genericpath: file /Users/m.a.sitnikov/devel/mypy-tests/venv/lib/python3.8/site-packages/mypy/typeshed/stdlib/genericpath.pyi
LOG:  Loaded graph with 25 nodes (0.028 sec)
LOG:  Found 2 SCCs; largest has 24 nodes
LOG:  Processing 1 queued fresh SCCs
LOG:  Processing SCC singleton (sample) as inherently stale
sample.py:5: error: Value of type variable "SupportsLessThanT" of "sorted" cannot be "Optional[str]"
LOG:  Deleting sample sample.py sample.meta.json sample.data.json
LOG:  No fresh SCCs left in queue
LOG:  Build finished in 0.130 seconds with 25 modules, and 2 errors
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: -v
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.8.2, 3.7.8
- Operating system and version: ProductName:	macOS, ProductVersion:	11.2.3, BuildVersion:	20D91

**Comment by pomo-mondreganto** [comment]
That's the expected behaviour as `NoneType` is not comparable to `str`, so `sorted(List[Optional[str]])` must fail the typecheck.

**Comment by erictraut** [comment]
You can work around this by changing it to:
```python
row.get(key, "")
```

The default value (`""`) will never be used in this case, but it provides a guarantee to the type checker that the set will never contain `None`.

**Comment by MihanixA** [comment]
@pomo-mondreganto 
> That's the expected behaviour as NoneType is not comparable to str, so sorted(List[Optional[str]]) must fail the typecheck.
True. Still the error message seems not so relevant

@erictraut, `row.get(key, "")` works fine.
I guess I minimized my code sample too much and accidentally removed if-condition in set comprehension 

```python3
from typing import Dict, List, Optional


def f(row: Dict[str, str], cols: List[str]) -> List[Optional[str]]:
    return list(sorted({row.get(key) for key in cols if row.get(key)}))
```

This code should not have problem comparing `NoneType` and `str` though it gets the same error

**Comment by erictraut** [comment]
A type checker has no way of knowing that `row.get(key)` is going to produce the same result every time. 

You would need to use one of these techniques:
```python
    return list(sorted({x for key in cols if (x := row.get(key))}))
```

or
```python
    return list(sorted({row.get(key, "") for key in cols if row.get(key)}))
```
